### PR TITLE
Task monitor refactoring

### PR DESCRIPTION
**Pull request by geofjamg**

### CODE REVIEW

File: afs/afs-ext-base/src/main/java/com/powsybl/afs/ext/base/LocalNetworkService.java [hunk 1]
```diff
@@ -116,20 +120,28 @@ private ModifiedNetwork loadNetworkFromProjectCase(ProjectCase projectCase) {
         }
     }
 
-    private <T extends ProjectFile & ProjectCase> ModifiedNetwork loadNetwork(T projectCase) {
+    private ModifiedNetwork loadNetwork(T projectCase) {
```
Comment (mathbagu) [inline afs/afs-ext-base/src/main/java/com/powsybl/afs/ext/base/LocalNetworkService.java]: This method should be static or remove the projectCase parameter to use the projectCase member.

File: afs/afs-ext-base/src/main/java/com/powsybl/afs/ext/base/NetworkService.java [hunk 2]
```diff
@@ -12,15 +12,15 @@
 /**
  * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
  */
-public interface NetworkService {
+public interface NetworkService<T extends ProjectFile & ProjectCase> {
```
[the remark this answers is not in the document]
Comment (mathbagu) [reply afs/afs-ext-base/src/main/java/com/powsybl/afs/ext/base/NetworkService.java]: Remove this generic parameter if possible

File: scripting/src/main/java/com/powsybl/scripting/RunScriptTool.java [hunk 3]
```diff
@@ -102,8 +102,10 @@ public void run(CommandLine line, ToolRunningContext context) throws Exception {
         Path file = context.getFileSystem().getPath(line.getOptionValue(FILE));
         Writer writer = new OutputStreamWriter(context.getOutputStream());
         try {
+            TaskMonitor taskMonitor = new LocalTaskMonitor();
+            taskMonitor.addListener(new SoutTaskListener(System.out));
```
[the remark this answers is not in the document]
Comment (mathbagu) [reply scripting/src/main/java/com/powsybl/scripting/RunScriptTool.java]: Use context.getOutputStream() instead

File: afs/afs-core/src/main/java/com/powsybl/afs/AppData.java [hunk 4]
```diff
@@ -201,13 +172,13 @@ public ListenableAppStorage getRemotelyAccessibleStorage(String fileSystemName)
...
 
-    <U> U findService(Class<U> serviceClass, boolean remote) {
+    <T extends ProjectFile, U> ServiceExtension<T, U> findService(Class<U> serviceClass, boolean remote) {
         ServiceExtension.ServiceKey<U> serviceKey = new ServiceExtension.ServiceKey<>(serviceClass, remote);
-        U service = (U) services.get(serviceKey);
-        if (service == null) {
-            throw new AfsException("No service found for key " + serviceKey + "");
+        ServiceExtension<T, U> serviceExtension = (ServiceExtension<T, U>) serviceExtensions.get(serviceKey);
+        if (serviceExtension == null) {
+            throw new AfsException("No service extension found for key " + serviceKey + "");
```
Comment (mathbagu) [inline afs/afs-core/src/main/java/com/powsybl/afs/AppData.java]: Remove the quotes at the end of this message

File: afs/afs-core/src/main/java/com/powsybl/afs/LocalTaskMonitor.java [hunk 5]
```diff
@@ -40,34 +47,39 @@ private void setRevision(long revision) {
...
+    public void stopTask(UUID id) {
+        Objects.requireNonNull(id);
         lock.lock();
         try {
             revision++;
-            tasks.remove(projectFile.getId());
+            ModifiableTask task = tasks.remove(id);
+            if (task == null) {
+                throw new IllegalArgumentException("Task '" + id + "' not found");
+            }
```
Comment (mathbagu) [inline afs/afs-core/src/main/java/com/powsybl/afs/LocalTaskMonitor.java]: You should probably increment the revision after this check

File: afs/afs-core/src/main/java/com/powsybl/afs/SoutTaskListener.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+    @Override
+    public String getProjectId() {
+        return null;
+    }
+
+    @Override
+    public void onEvent(TaskEvent event) {
+        if (event instanceof UpdateTaskMessageEvent) {
+            out.println(((UpdateTaskMessageEvent) event).getMessage());
```
Comment (mathbagu) [inline afs/afs-core/src/main/java/com/powsybl/afs/SoutTaskListener.java]: Maybe you should also print the taskID

File: afs/afs-core/src/main/java/com/powsybl/afs/TaskMonitor.java [hunk 7]
```diff
@@ -16,31 +17,39 @@
...
 
         protected String message;
 
         protected long revision;
 
-        protected Task(ProjectFile projectFile, String message, long revision) {
-            this.projectFile = Objects.requireNonNull(projectFile);
+        protected Task(String name, String message, long revision) {
+            id = UUID.randomUUID();
+            this.name = name;
```
Comment (mathbagu) [inline afs/afs-core/src/main/java/com/powsybl/afs/TaskMonitor.java]: Is null allowed here?

File: afs/afs-ext-base/src/main/java/com/powsybl/afs/ext/base/LocalNetworkService.java [hunk 8]
```diff
@@ -116,20 +120,28 @@ private ModifiedNetwork loadNetworkFromProjectCase(ProjectCase projectCase) {
...
+            });
         } catch (ExecutionException e) {
             throw new UncheckedExecutionException(e);
         }
     }
 
     @Override
-    public <T extends ProjectFile & ProjectCase> String queryNetwork(T projectCase, String groovyScript) {
+    public String queryNetwork(String groovyScript) {
         Objects.requireNonNull(projectCase);
```
Comment (mathbagu) [inline afs/afs-ext-base/src/main/java/com/powsybl/afs/ext/base/LocalNetworkService.java]: Remove this check

File: afs/afs-ext-base/src/main/java/com/powsybl/afs/ext/base/LocalNetworkService.java (hunk 1 above)
Comment (geofjamg) [reply afs/afs-ext-base/src/main/java/com/powsybl/afs/ext/base/LocalNetworkService.java]: Done

File: afs/afs-core/src/main/java/com/powsybl/afs/AppData.java (hunk 4 above)
Comment (geofjamg) [reply afs/afs-core/src/main/java/com/powsybl/afs/AppData.java]: Done

File: afs/afs-core/src/main/java/com/powsybl/afs/LocalTaskMonitor.java (hunk 5 above)
Comment (geofjamg) [reply afs/afs-core/src/main/java/com/powsybl/afs/LocalTaskMonitor.java]: Right

File: afs/afs-core/src/main/java/com/powsybl/afs/SoutTaskListener.java (hunk 6 above)
Comment (geofjamg) [reply afs/afs-core/src/main/java/com/powsybl/afs/SoutTaskListener.java]: This is not a usefull user info

File: afs/afs-core/src/main/java/com/powsybl/afs/TaskMonitor.java (hunk 7 above)
Comment (geofjamg) [reply afs/afs-core/src/main/java/com/powsybl/afs/TaskMonitor.java]: No, check added

File: afs/afs-ext-base/src/main/java/com/powsybl/afs/ext/base/LocalNetworkService.java (hunk 8 above)
Comment (geofjamg) [reply afs/afs-ext-base/src/main/java/com/powsybl/afs/ext/base/LocalNetworkService.java]: Done

File: afs/afs-ext-base/src/main/java/com/powsybl/afs/ext/base/NetworkService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (geofjamg) [reply afs/afs-ext-base/src/main/java/com/powsybl/afs/ext/base/NetworkService.java]: Done

File: scripting/src/main/java/com/powsybl/scripting/RunScriptTool.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (geofjamg) [reply scripting/src/main/java/com/powsybl/scripting/RunScriptTool.java]: Done

**Pull request merged**